core/src/test/java/com/linecorp/armeria/client/endpoint/GradualStartEndpointSelectionStrategyTest.java

```diff
@@ -0,0 +1,135 @@
+        final String groupName = "weighted";
+
+        EndpointGroupRegistry.register(groupName, endpointGroup, strategy);
+
+        assertThat(EndpointGroupRegistry.selectNode(ctx, groupName).authority()).isEqualTo("127.0.0.1:1234");
+        assertThat(EndpointGroupRegistry.selectNode(ctx, groupName).authority()).isEqualTo("127.0.0.1:2345");
+        assertThat(EndpointGroupRegistry.selectNode(ctx, groupName).authority()).isEqualTo("127.0.0.1:3456");
+        assertThat(EndpointGroupRegistry.selectNode(ctx, groupName).authority()).isEqualTo("127.0.0.1:2345");
+        assertThat(EndpointGroupRegistry.selectNode(ctx, groupName).authority()).isEqualTo("127.0.0.1:3456");
+        assertThat(EndpointGroupRegistry.selectNode(ctx, groupName).authority()).isEqualTo("127.0.0.1:3456");
+    }
+
+    @Test
+    void selectFromDynamicEndpointGroup() throws Exception {
+        final AtomicInteger counter = new AtomicInteger();
+        final EndpointWeightTransition transition = (e, step, maxStep) -> {
+            counter.incrementAndGet();
+            return EndpointWeightTransition.linear().compute(e, step, maxStep);
+        };
+        final EndpointSelectionStrategy strategy = GradualStartEndpointSelectionStrategy
+                .builder()
+                .weightTransition(transition)
+                .slowStartInterval(Duration.ofSeconds(1))
+                .numberOfSteps(10)
+                .build();
+
+        final TestDynamicEndpointGroup endpointGroup = new TestDynamicEndpointGroup();
+        EndpointGroupRegistry.register("dynamic", endpointGroup, strategy);
+        endpointGroup.updateEndpoints(ImmutableList.of(Endpoint.of("127.0.0.1", 1000)));
+
+        final EndpointSelector selector = EndpointGroupRegistry.getNodeSelector("dynamic");
+        // Schedule updateEndpointWeight.
+        TimeUnit.SECONDS.sleep(1);
+        assertThat(selector.select(ctx)).isEqualTo(Endpoint.of("127.0.0.1", 1000));
+
+        endpointGroup.updateEndpoints(ImmutableList.of(
+                Endpoint.of("127.0.0.1", 1111).withWeight(100)));
+
+        await().atMost(Duration.ofSeconds(1))
+               .untilAsserted(() -> {
+                   final Endpoint e = selector.select(ctx);
+                   assertThat(e.host()).isEqualTo("127.0.0.1");
+                   assertThat(e.port()).isEqualTo(1111);
+                   assertThat(e.weight()).isGreaterThanOrEqualTo(1);
+               });
+        await().atMost(Duration.ofSeconds(1))
+               .untilAsserted(() -> assertThat(selector.select(ctx).weight()).isGreaterThanOrEqualTo(10));
+        await().atMost(Duration.ofSeconds(1))
+               .untilAsserted(() -> assertThat(selector.select(ctx).weight()).isGreaterThanOrEqualTo(20));
+        await().atMost(Duration.ofSeconds(1))
+               .untilAsserted(() -> assertThat(selector.select(ctx).weight()).isGreaterThanOrEqualTo(30));
+        // Set known endpoint but the weight continue to rump up with previous weight.
```

Comment:
- to rump -> ramping
- the weight continue -> the weight should continue

What would happen if the new endpoint has the same host and port but with different weight? Could you add a test case for it as well?